classes/ActionScheduler_wpPostStore.php

```diff
@@ -462,25 +465,101 @@ protected function generate_claim_id() {
 	}
 
 	/**
-	 * @param string $claim_id
-	 * @param int $limit
+	 * @param string   $claim_id
+	 * @param int      $limit
 	 * @param DateTime $before_date Should use UTC timezone.
+	 * @param string   $group       Claim only actions in the given group.
+	 *
 	 * @return int The number of actions that were claimed
 	 * @throws RuntimeException
 	 */
-	protected function claim_actions( $claim_id, $limit, DateTime $before_date = NULL ) {
+	protected function claim_actions( $claim_id, $limit, DateTime $before_date = null, $group = '' ) {
 		/** @var wpdb $wpdb */
 		global $wpdb;
 
-		$date = is_null($before_date) ? as_get_datetime_object() : clone $before_date;
-		// can't use $wpdb->update() because of the <= condition, using post_modified to take advantage of indexes
-		$sql = "UPDATE {$wpdb->posts} SET post_password = %s, post_modified_gmt = %s, post_modified = %s WHERE post_type = %s AND post_status = %s AND post_password = '' AND post_date_gmt <= %s ORDER BY menu_order ASC, post_date_gmt ASC, ID ASC LIMIT %d";
-		$sql = $wpdb->prepare( $sql, array( $claim_id, current_time('mysql', true), current_time('mysql'), self::POST_TYPE, 'pending', $date->format('Y-m-d H:i:s'), $limit ) );
-		$rows_affected = $wpdb->query($sql);
+		$date = null === $before_date ? as_get_datetime_object() : clone $before_date;
+
+		// Set up a query for post IDs to use later.
+		$query      = new WP_Query();
+		$query_args = array(
+			'fields'           => 'ids',
+			'post_type'        => self::POST_TYPE,
+			'post_status'      => ActionScheduler_Store::STATUS_PENDING,
+			'has_password'     => false,
+			'posts_per_page'   => $limit,
```

Comment:
Because this select and the claim are being run in separate queries, there is a race condition where:

1. this query returns a valid set of IDs that are due to be claimed
1. another process claims that same set of results, because it ran the same query with the same order by clause
1. the claim query is run below, but doesn't stake anything because `post_password` is no longer `''`

That's not a critical issue. It just means this attempt to stake a claim will be ignored.

However, I think it can be avoided quite easily by increasing the `'posts_per_page'` value here to something like `2 * $limit` OR `3 * $limit`. That would allow 2 or 3 processes to be attempting to stake claims at the same time. Because of the ease of addressing it, IMO it's worth doing.

The `$limit` is necessary of course to avoid the MySQL query below becoming too long and running into things like WP Engine's [1024 character limit on MySQL queries](https://wpengine.com/support/troubleshoot-wordpress-wp-engine-error-log/#longquery). But I think even with a 2x or 3x increase in the limit, we'll generally fall safely within that with default configurations.

Reply:
This makes sense. I think `3 * $limit` is reasonable.